scalikejdbc-streams/src/main/scala/scalikejdbc/streams/ExtractedResultIterator.scala

```diff
@@ -0,0 +1,55 @@
+package scalikejdbc.streams
+
+import java.sql.ResultSet
+
+import scalikejdbc.{ ResultSetCursor, WrappedResultSet }
+
+abstract class ExtractedResultIterator[+A](rs: ResultSet, autoClose: Boolean)(extract: WrappedResultSet => A) extends BufferedIterator[A] with CloseableIterator[A] {
+  private[this] var state = 0 // 0: no data, 1: cached, 2: finished
+  private[this] var cached: A = null.asInstanceOf[A]
+
+  protected[this] final def finished(): A = {
+    state = 2
+    null.asInstanceOf[A]
+  }
+
+  def head: A = {
+    update()
+    if (state == 1) cached
+    else throw new NoSuchElementException("head on empty iterator")
+  }
+
+  private[this] def update(): Unit = {
+    if (state == 0) {
```

Comment:
`private[this] var` is compiled to direct field access, contribute for (a few) performance matter, isn't it? how about this:

```scala
  @inline
  private[this] def isNoDataState: Boolean = state == 0

  @inline
  private[this] def isCachedState: Boolean = state == 1

  @inline
  private[this] def markedAsNoData(): Unit = state = 0

  @inline
  private[this] def markedAsCached(): Unit = state = 1
```